README.md
# Pokigen

## Description
A simple CNN Beta-VAE based Pokémon image generator written in simplified Tensorflow 2.3 Keras.
The CNN beta-VAE structure features a 
- Variational Convolution based Encoder encoding the images in a latent space as a Gaussian distribution consisting mainly in encoding blocks with different parameters depending on the version of the network you are using.
- Convolutional based decoder with upscaling decoding images from latent space back to normale images.

We present 3 varients of the network :
  - Light version : a simple CNN maxpooling, cnn upsampling architecture
  - Mobilenet : uses the convolutional layers of mobile net in encoder and cnn upsampling layers for decoding
  - Full : a fully fledged CNN encoder decoder with deep resnet structure. Needs alot of time and power to learn

The input of the autoencoder is a 128X128X3 image
The output is also a 128X128X3 image

# Objective
As for today, the software is a simple beta vae based pokemon images generator. The real objective is to correlate the pokemon name with its photo and to be able to generate pokemons from the name.




Changelog.txt
Changelog

V1.3
------
A more stable training by shifting from sigma to logvar in latent 
space gaussian distribution samplig
A more factorized coding style

V1.2
------
Enhanced Network structure
Provided multiple versions of the network including a mobile net transfer learning based version
V 1.1
------
Upgraded code to simplify the operations

V 1.0
------
First run
Added cnn_beta_vae version 1
Added training code


src/train.py
"""
Training code for Pokigen application
Licence : MIT
Author  : Saifeddine ALOUI
Description :
    Uses a database of images to build an image generator
    This code is built to use the cnnbetavae network to learn how to generate pokemons


"""
import argparse
import os

# Preprare to parse arguments
parser = argparse.ArgumentParser()
parser.add_argument("-db","--database", help="Database folder containing the images to be used to learn")
parser.add_argument("-m","--model_weights_file", help="Link to the model weights file (useful to resume learning and to put checkpoints)")
parser.add_argument("-r","--restart", action="store_true", help="if set the model will be reset, otherwise last checkpoint will be loaded")
parser.add_argument("-e","--epochs", type=int, help="The number of epochs for training")
parser.add_argument("-n","--net", help="The network to be used training")

parser.add_argument("-l","--latent_size", type=int, help="Latent space size (default to 1)")
parser.add_argument("-b","--beta", type=float, help="The value for beta (default 0.001)")

args = parser.parse_args()

# Set default arguments
if args.database is None:
    args.database = "database"

if args.epochs is None:
    args.epochs = 500

if args.model_weights_file is None:
    args.model_weights_file = "models/model_weights.h5"

if args.beta is None:
    args.beta = 2
    
if args.latent_size is None:
    args.latent_size = 27
    
if args.net is None:
    #print("Using full network")
    #from cnn_beta_vae import CNNBetaVAE
    #print("Using mobilenet based network")
    #from cnn_beta_vae_mobilenet import CNNBetaVAE
    print("Using light network")
    from cnn_beta_vae_light import CNNBetaVAE
else:
    if args.net=="cnn_beta_vae":
        print("Using full network")
        from cnn_beta_vae import CNNBetaVAE
    elif args.net=="cnn_beta_vae_mobilenet":
        print("Using mobilenet network")
        from cnn_beta_vae_mobilenet import CNNBetaVAE
    else: 
        print("Using light network")
        from cnn_beta_vae_light import CNNBetaVAE
args.restart = True
# Do the training process
def main():
    """
    Main code to train the Pokigen CNN Beta VAE network
    """
    cnnbvae = CNNBetaVAE((128,128,3),args.latent_size,args.beta,model_weights_path=args.model_weights_file,use_logvar=True)

    # Check if the weights file exists and that the reset option was not issued
    if os.path.exists(args.model_weights_file) and not args.restart:
        cnnbvae.load_weights(args.model_weights_file)
    

    # Load training/validation database and preprocess it
    images = cnnbvae.preprocess_images(cnnbvae.load_images(args.database))

    # Train the network
    cnnbvae.learn(images, epochs=args.epochs)

    # Plot loss
    plt.figure()
    plt.grid()
    plt.plot(cnnbvae.loss_buffer, label="Loss")
    plt.plot(cnnbvae.val_loss_buffer, label="Validation Loss")
    plt.title("Loss curves")
    plt.savefig("../figures/{}_loss.png".format(args.net))

if __name__ == "__main__":
    # execute only if run as a script
    main()


src/cnn_beta_vae.py
"""
CNN Beta Vae Net 128
Author  : Saifeddine ALOUI
Licence : MIT
Description :
    This is a simple CNN based beta Variational auto encoder for
    128X128X3 image generation.

    The RootCNNBetaVAE class should be inherited and encoder and decoder models
    should be defined. Use the children classes for your application

    The variational autoencoder encodes the images in a latent space as
    a random variable with gaussian distribution then decodes them back

    The encoder goes from 128X128X3 to a list of [N,N,N] dimensions [Mean, Var, sample]
    The decoder goes from N to 128X128X3 (it takes the generated sample and generate the output image)

    First build the CNNBetaVAE128 object by passing in the
    To learn, you only need to call the learn method and pass the NX128X128X3 numpy array
    containing the images to learn from.

    To generate the images, you can sample from the latent space
"""

# OS useful to load files etc
import os

# Numpy to do array manipulation
import numpy as np

# Tensorflow/Keras layers, models, backend and callbacks
import tensorflow
from tensorflow.keras.layers import Dense, Conv2D, Conv2DTranspose, Input, MaxPool2D, Flatten, UpSampling2D, Reshape, Lambda, Concatenate, BatchNormalization, Add
from tensorflow.keras.activations import sigmoid, relu
from tensorflow.keras.models import Model
import tensorflow.keras.backend as K
import tensorflow as tf
from tensorflow.keras.callbacks import Callback, EarlyStopping, ModelCheckpoint
tf.compat.v1.disable_eager_execution()

# Open Cv for image manipulation and preproicessing
import cv2


class RootCNNBetaVAE(Callback):
    """
    CNN Beta Variational Encoder
    """
    def __init__(self, image_shape, latent_size, beta, model_weights_path = "model.h5", use_logvar=False):
        """
        Builds the model
        image_shape : The input Image Shape for example for an RGB picture of 128X128
                      you put (128,128,3)
        latent__size : The size of latent space  example 128 or even 1 for binary images generation
        beta : The weight to apply on the KL divergence (like in the beta vae paper)
        """
        # Save parameters
        self.reset_buffers()
        self.image_shape = image_shape
        self.latent_size = latent_size
        self.beta = beta
        self.use_logvar =use_logvar
        self.model_weights_path = model_weights_path
    
    def loss(self, y_true, y_pred):
        """
        Loss function
        Composed of a weighted sum between two losses
        reconstruction loss : to ensure tha capability of the network to reproduce the images
        kl loss : to force the latent space to have a tight distribution 
        """
        reconstruction_loss = K.mean(K.square(y_true-y_pred))
        if self.use_logvar :
            kl_loss = - 0.5 * K.mean(1 + self.encoded_sig - K.square(self.encoded_mean) - K.exp(self.encoded_sig), axis=-1)
        else:
            kl_loss = -0.5 * K.mean(1 + K.log(K.square(self.encoded_sig)) - tf.square(self.encoded_sig) - tf.square(self.encoded_mean))

        return reconstruction_loss + self.beta* kl_loss


    def encoderProcessingBlock(self, inp, nb_kernels, kernelsize=(3,3), reduction_factor=1, activation=relu, batch_normalize=False, is_residual=False):
        """
        An encoder processing block
        inp : Input to be processed by the layer
        nb_kernels : The number of kernels to apply in convolutional layer
        kernel_size : The size of the kernel to be used in convolutional layer
        reduction_factor : maxpooling factor (if 1 or less, maxpooling is not applyed)
        activation : The activation function to be applied
        batch_normalize : A boolean to tell if a batch normalization should be applied before activation
        is_residual : A boolean to tell if this is a risidual block. Warning!! If True make sure that the 
                      input and output sizes are exactly the same (no pooling and must have the same number of channels) 
        """
        x = Conv2D(nb_kernels,kernelsize, padding="same")(inp)
        if batch_normalize:
            x = BatchNormalization()(x)
        x = activation(x)
        if is_residual:
            x = Add()([inp,x])

        if reduction_factor>1:
            x = MaxPool2D((reduction_factor,reduction_factor))(x)
        return x

    def decoderProcessingBlock(self, inp, nb_kernels, kernelsize=(3,3), upscaling_factor=1, activation=relu, batch_normalize=False, is_residual=False):
        """
        A decoder processing block
        inp : Input to be processed by the layer
        nb_kernels : The number of kernels to apply in convolutional layer
        kernel_size : The size of the kernel to be used in convolutional layer
        upscaling_factor : upscaling factor (if 1 or less, upscaling is not applyed)
        activation : The activation function to be applied
        batch_normalize : A boolean to tell if a batch normalization should be applied before activation
        is_residual : A boolean to tell if this is a risidual block. Warning!! If True make sure that the 
                      input and output sizes are exactly the same (no pooling and must have the same number of channels) 
        """
        x = Conv2D(nb_kernels,kernelsize, padding="same")(inp)
        if batch_normalize:
            x = BatchNormalization()(x)
        x = activation(x)
        if is_residual:
            x = Add()([inp,x])
        if upscaling_factor>1:
            x = UpSampling2D((upscaling_factor,upscaling_factor))(x)
        return x


    # ============ Weights loading/storing methods ==================
    def load_weights(self, model_weights_file):
        """
        Loads a weights file
        (without the right weights the network is useless and need
        to be retrained)
        model_weights_file : The file containing the weights to be used
        """
        self.bvae.load_weights(model_weights_file)

    def save_weights(self, model_weights_file):
        """
        Saves weights to a weights file
        model_weights_file : The file in which to store the weights
        """
        self.bvae.load_weights(model_weights_file)

    # =========  Data manipulation methods ==============

    def load_images(self, database_path, return_image_file_names=False):
        """
        Loads the database of images and reshape them according to the needed format
        database_path.
        
        dataase_path : The path to the folder containing database of images to be loaded
        return : A numpy array of size batch, width, height, channels containing a stacking of all loaded image files
        """

        image_files = sorted([f for f in os.listdir(database_path) if f.lower().endswith(".png") or  f.lower().endswith(".jpg")])
        images=[]
        for pokemons_file in image_files:
            img = cv2.imread(os.path.join(database_path, pokemons_file), cv2.IMREAD_UNCHANGED)
            if len(img.shape)==4:
                alpha_channel = img[:, :, 3]
                _, mask = cv2.threshold(alpha_channel, 254, 255, cv2.THRESH_BINARY)  # binarize mask
                color = img[:, :, :3]
                img = cv2.bitwise_not(cv2.bitwise_not(color, mask=not mask))
                img = cv2.cvtColor(img, cv2.COLOR_BGRA2RGB)
            else:
                img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
            images.append(np.reshape(cv2.resize(img,(self.image_shape[0],self.image_shape[1])),(1,self.image_shape[0],self.image_shape[1],self.image_shape[2])))
        images=(np.vstack(images))

        if return_image_file_names:
            return images, image_files
        else:
            return images

    def preprocess_images(self, images):
        """
        Preprocesses images to prepare them to be used by the network
        the operation assumes the images were loaded using the load_images method that prepares
        the images and resizes them to have the right format.

        The images are in the 0-255 RGB format

        this operation will normalize their values to be between -1 and 1 (useful since we use tanh as activation function
        in all layers)

        images: A 4D vector in form : Batches, width, Height, Chanels (RGB)

        returns : The normalized images numpy array
        """
        # Normalize the images to be between -1 and 1
        images=images.astype(np.float32)/255.0
        return images

    def postprocess_outputs(self, outputs):
        """
        Postprocesses the outputsto be usable as 
        regular 255 range RGB images 
        """
        outputs=outputs*255
        return outputs

    # =================== Reparameterization Trick !! It's a VAE ================
    def sampling(self, args):
        """
        Reparameterization trick by sampling for an isotropic unit Gaussian.
        To be able to back propagate through a sampling step, this trick consists
        in changing Q(z|x) to be z_mean+z_sig*N(0,1) or z_mean+exp(0.5*z_logvar)*N(0,1)
        depending on wether you use standard deviation or logvariance when computing
        the statistics in your latent space 
        # Arguments
            args : (z_mean,z_sig)
                    z_mean     : mean of Q(z|X)
                    z_sig      : standard deviation or (log of variance if self.use_logvar=True) of Q(z|X)

        # Returns
            z (tensor): sampled latent vector
        """
        z_mean, z_sig = args
        batch = K.shape(z_mean)[0]
        dim = K.int_shape(z_mean)[1]
        # by default, random_normal has mean=0 and std=1.0
        epsilon = K.random_normal(shape=(batch, dim))
        if self.use_logvar:
            z_log_var = z_sig
            return  z_mean + K.exp(0.5 * z_log_var) * epsilon
        else:
            return  z_mean + z_sig * epsilon        


    # ========== Network Learning and predicting ==================
    def learn(self, images, epochs=500, validation_split=0.2):
        # Fit unsupervised
        print("Fitting auto encoder")
        self.bvae.fit(
            images,
            images, 
            validation_split=validation_split,  
            epochs=epochs, 
            callbacks=[
                self, 
                EarlyStopping(patience=200), 
                ModelCheckpoint(self.model_weights_path,save_best_only=True)
                ]
        )

    def predict(self, landmarks):
        return self.bvae.predict(landmarks)

    def convert(self, image):
        decoded = self.bvae.predict(image)
        return decoded

    # =========== Learning Callbacks ================================ 
    def reset_buffers(self):
        """
        Resets the statistics buffers
        """
        self.loss_buffer=[]
        self.val_loss_buffer=[]
        self.learning_encoding_buffer=[]
        self.epoch=0

    def on_epoch_end(self, epoch, logs=None):
        self.epoch=epoch
        self.loss_buffer.append(logs["loss"])
        self.val_loss_buffer.append(logs["val_loss"])



"""
        # Build encoder
        input_data = Input(image_shape,name="input")
        x = self.encoderProcessingBlock(input_data, 64)
        x = self.encoderProcessingBlock(x, 128)
        x = Flatten()(x)
        x = Dense(16*16,activation="tanh")(x)
        # Reparametrization trick
        self.encoded_mean = Dense(latent_size)(x)
        self.encoded_sig = Dense(latent_size)(x)
        self.encoded_Data = Lambda(self.sampling, output_shape=(latent_size,), name="Encoded")([self.encoded_mean, self.encoded_sig])
        
        # Decoder
        latent_input=Input((latent_size))
        x = Dense(16*16,activation="tanh")(latent_input)
        x = Reshape((16,16,1))(x)
        x = self.decoderProcessingBlock(x,32, upscaling_factor=2)
        x_r = self.decoderProcessingBlock(x,64, upscaling_factor=2)
        x_g = self.decoderProcessingBlock(x,64, upscaling_factor=2)
        x_b = self.decoderProcessingBlock(x,64, upscaling_factor=2)

        x_r = self.decoderProcessingBlock(x_r,128, upscaling_factor=2)
        x_g = self.decoderProcessingBlock(x_g,128, upscaling_factor=2)
        x_b = self.decoderProcessingBlock(x_b,128, upscaling_factor=2)
        
        x_r = self.decoderProcessingBlock(x_r,1, upscaling_factor=1)
        x_g = self.decoderProcessingBlock(x_g,1, upscaling_factor=1)
        x_b = self.decoderProcessingBlock(x_b,1, upscaling_factor=1)
        decoded = Concatenate(axis=3)([x_r,x_g,x_b])

"""


src/cnn_beta_vae_light.py
"""
CNN Beta Vae Net 128
Author  : Saifeddine ALOUI
Licence : MIT
Description :
    This is a simple CNN based beta Variational auto encoder for
    128X128X3 image generation

    The variational autoencoder encodes the images in a latent space as
    a random variable with gaussian distribution then decodes them back

    The encoder goes from 128X128X3 to a list of [N,N,N] dimensions [Mean, Var, sample]
    The decoder goes from N to 128X128X3 (it takes the generated sample and generate the output image)

    First build the CNNBetaVAE128 object by passing in the
    To learn, you only need to call the learn method and pass the NX128X128X3 numpy array
    containing the images to learn from.

    To generate the images, you can sample from the latent space
"""

# OS useful to load files etc
import os

# Numpy to do array manipulation
import numpy as np

# Tensorflow/Keras layers, models, backend and callbacks
import tensorflow
from tensorflow.keras.layers import Dense, Conv2D, Conv2DTranspose, Input, MaxPool2D, Flatten, UpSampling2D, Reshape, Lambda, Concatenate, BatchNormalization, Add
from tensorflow.keras.activations import sigmoid, relu
from tensorflow.keras.models import Model
import tensorflow.keras.backend as K
import tensorflow as tf
from tensorflow.keras.callbacks import Callback, EarlyStopping, ModelCheckpoint
tf.compat.v1.disable_eager_execution()

from cnn_beta_vae import RootCNNBetaVAE
# Open Cv for image manipulation and preproicessing
import cv2


class CNNBetaVAE(RootCNNBetaVAE):
    """
    CNN Beta Variational Encoder 128
    """
    def __init__(self, image_shape, latent_size, beta, model_weights_path = "model.h5", use_logvar=False):
        """
        Builds the model
        image_shape : The input Image Shape for example for an RGB picture of 128X128
                      you put (128,128,3)
        latent__size : The size of latent space  example 128 or even 1 for binary images generation
        beta : The weight to apply on the KL divergence (like in the beta vae paper)
        """
        # Save parameters
        RootCNNBetaVAE.__init__(self, image_shape, latent_size, beta, model_weights_path, use_logvar)
            
        # Build encoder
        input_data = Input(image_shape,name="input")
        x = self.encoderProcessingBlock(input_data, 3, reduction_factor=2)
        x = self.encoderProcessingBlock(x, 16, reduction_factor=2)
        x = self.encoderProcessingBlock(x, 32, reduction_factor=2)
        x = self.encoderProcessingBlock(x, 64, reduction_factor=2)
        x = Flatten()(x)
        x = Dense(4*4,activation="tanh")(x)
        # Reparametrization trick
        self.encoded_mean = Dense(latent_size)(x)
        self.encoded_sig = Dense(latent_size)(x)
        self.encoded_Data = Lambda(self.sampling, output_shape=(latent_size,), name="Encoded")([self.encoded_mean, self.encoded_sig])

        # Decoder
        latent_input=Input((latent_size))
        x = Dense(4*4,activation="tanh")(latent_input)
        x = Reshape((4,4,1))(x)
        x = self.decoderProcessingBlock(x, 8, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 8, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 8, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 8, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 8, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 512)
        
        x = self.decoderProcessingBlock(x, 3)
        decoded = self.decoderProcessingBlock(x, 3)
        

        # Build models
        self.encoder=Model(input_data,[self.encoded_mean,self.encoded_sig,self.encoded_Data], name="encoder")
        self.decoder=Model(latent_input,decoded, name="decoder")
        self.bvae=Model(input_data,self.decoder(self.encoder(input_data)[2]), name="bvae")

        # Compile the models and show some summaries
        self.bvae.compile(loss=self.loss)

        # Show the three main components the encoder, the decoder and the full VAE
        self.encoder.summary()
        self.decoder.summary()
        self.bvae.summary()
    

"""
        # Build encoder
        input_data = Input(image_shape,name="input")
        x = self.encoderProcessingBlock(input_data, 64)
        x = self.encoderProcessingBlock(x, 128)
        x = Flatten()(x)
        x = Dense(16*16,activation="tanh")(x)
        # Reparametrization trick
        self.encoded_mean = Dense(latent_size)(x)
        self.encoded_sig = Dense(latent_size)(x)
        self.encoded_Data = Lambda(self.sampling, output_shape=(latent_size,), name="Encoded")([self.encoded_mean, self.encoded_sig])
        
        # Decoder
        latent_input=Input((latent_size))
        x = Dense(16*16,activation="tanh")(latent_input)
        x = Reshape((16,16,1))(x)
        x = self.decoderProcessingBlock(x,32, upscaling_factor=2)
        x_r = self.decoderProcessingBlock(x,64, upscaling_factor=2)
        x_g = self.decoderProcessingBlock(x,64, upscaling_factor=2)
        x_b = self.decoderProcessingBlock(x,64, upscaling_factor=2)

        x_r = self.decoderProcessingBlock(x_r,128, upscaling_factor=2)
        x_g = self.decoderProcessingBlock(x_g,128, upscaling_factor=2)
        x_b = self.decoderProcessingBlock(x_b,128, upscaling_factor=2)
        
        x_r = self.decoderProcessingBlock(x_r,1, upscaling_factor=1)
        x_g = self.decoderProcessingBlock(x_g,1, upscaling_factor=1)
        x_b = self.decoderProcessingBlock(x_b,1, upscaling_factor=1)
        decoded = Concatenate(axis=3)([x_r,x_g,x_b])

"""


src/cnn_beta_vae_mobilenet.py
"""
CNN Beta Vae Net 128
Author  : Saifeddine ALOUI
Licence : MIT
Description :
    This is a simple CNN based beta Variational auto encoder for
    128X128X3 image generation

    The variational autoencoder encodes the images in a latent space as
    a random variable with gaussian distribution then decodes them back

    The encoder goes from 128X128X3 to a list of [N,N,N] dimensions [Mean, Var, sample]
    The decoder goes from N to 128X128X3 (it takes the generated sample and generate the output image)

    First build the CNNBetaVAE128 object by passing in the
    To learn, you only need to call the learn method and pass the NX128X128X3 numpy array
    containing the images to learn from.

    To generate the images, you can sample from the latent space
"""

# OS useful to load files etc
import os

# Numpy to do array manipulation
import numpy as np

# Tensorflow/Keras layers, models, backend and callbacks
import tensorflow
from tensorflow.keras.layers import Dense, Conv2D, Conv2DTranspose, Input, MaxPool2D, Flatten, UpSampling2D, Reshape, Lambda, Concatenate, BatchNormalization, Add
from tensorflow.keras.activations import sigmoid, relu
from tensorflow.keras.models import Model
import tensorflow.keras.backend as K
import tensorflow as tf
from tensorflow.keras.callbacks import Callback, EarlyStopping, ModelCheckpoint
tf.compat.v1.disable_eager_execution()

from cnn_beta_vae import RootCNNBetaVAE
# Open Cv for image manipulation and preproicessing
import cv2


class CNNBetaVAE(RootCNNBetaVAE):
    """
    CNN Beta Variational Encoder 128
    """
    def __init__(self, image_shape, latent_size, beta, model_weights_path = "model.h5", use_logvar=False):
        """
        Builds the model
        image_shape : The input Image Shape for example for an RGB picture of 128X128
                      you put (128,128,3)
        latent__size : The size of latent space  example 128 or even 1 for binary images generation
        beta : The weight to apply on the KL divergence (like in the beta vae paper)
        """
        # Save parameters
        RootCNNBetaVAE.__init__(self, image_shape, latent_size, beta, model_weights_path, use_logvar)

        # Recover mobilenet
        base_model = tensorflow.keras.applications.MobileNet(
            weights='imagenet',  # Load weights pre-trained on ImageNet.
            input_shape=image_shape,
            include_top=False)  # Do not include the ImageNet classifier at the top.
        # Freese base model
        base_model.trainable=False
            
        # Build encoder
        input_data = Input(image_shape,name="input")
        x= base_model(input_data)
        x = Flatten()(x)
        x = Dense(4*4,activation="tanh")(x)
        # Reparametrization trick
        self.encoded_mean = Dense(latent_size)(x)
        self.encoded_sig = Dense(latent_size)(x)
        self.encoded_Data = Lambda(self.sampling, output_shape=(latent_size,), name="Encoded")([self.encoded_mean, self.encoded_sig])

        # Decoder
        latent_input=Input((latent_size))
        x = Dense(4*4,activation="tanh")(latent_input)
        x = Reshape((4,4,1))(x)
        x = self.decoderProcessingBlock(x, 128, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 64, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 32, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 16, upscaling_factor=2)
        x = self.decoderProcessingBlock(x, 3, upscaling_factor=2)
        decoded = self.decoderProcessingBlock(x, 3)

        # Build models
        self.encoder=Model(input_data,[self.encoded_mean,self.encoded_sig,self.encoded_Data], name="encoder")
        self.decoder=Model(latent_input,decoded, name="decoder")
        self.bvae=Model(input_data,self.decoder(self.encoder(input_data)[2]), name="bvae")

        # Compile the models and show some summaries
        self.bvae.compile(loss=self.loss)

        # Show the three main components the encoder, the decoder and the full VAE
        self.encoder.summary()
        self.decoder.summary()
        self.bvae.summary()
    


src/cnn_beta_vae_full.py
"""
CNN Beta Vae Net 128
Author  : Saifeddine ALOUI
Licence : MIT
Description :
    This is a simple CNN based beta Variational auto encoder for
    128X128X3 image generation

    The variational autoencoder encodes the images in a latent space as
    a random variable with gaussian distribution then decodes them back

    The encoder goes from 128X128X3 to a list of [N,N,N] dimensions [Mean, Var, sample]
    The decoder goes from N to 128X128X3 (it takes the generated sample and generate the output image)

    First build the CNNBetaVAE128 object by passing in the
    To learn, you only need to call the learn method and pass the NX128X128X3 numpy array
    containing the images to learn from.

    To generate the images, you can sample from the latent space
"""

# OS useful to load files etc
import os

# Numpy to do array manipulation
import numpy as np

# Tensorflow/Keras layers, models, backend and callbacks
import tensorflow
from tensorflow.keras.layers import Dense, Conv2D, Conv2DTranspose, Input, MaxPool2D, Flatten, UpSampling2D, Reshape, Lambda, Concatenate, BatchNormalization, Add
from tensorflow.keras.activations import sigmoid, relu
from tensorflow.keras.models import Model
import tensorflow.keras.backend as K
import tensorflow as tf
from tensorflow.keras.callbacks import Callback, EarlyStopping, ModelCheckpoint
tf.compat.v1.disable_eager_execution()

from cnn_beta_vae import RootCNNBetaVAE
# Open Cv for image manipulation and preproicessing
import cv2


class CNNBetaVAE(RootCNNBetaVAE):
    """
    CNN Beta Variational Encoder 128
    """
    def __init__(self, image_shape, latent_size, beta, model_weights_path = "model.h5", use_logvar=False):
        """
        Builds the model
        image_shape : The input Image Shape for example for an RGB picture of 128X128
                      you put (128,128,3)
        latent__size : The size of latent space  example 128 or even 1 for binary images generation
        beta : The weight to apply on the KL divergence (like in the beta vae paper)
        """
        # Save parameters
        RootCNNBetaVAE.__init__(self, image_shape, latent_size, beta, model_weights_path, use_logvar)
            
        # Build encoder
        input_data = Input(image_shape,name="input")
        x = self.encoderProcessingBlock(input_data, 3)
        x = self.encoderProcessingBlock(x, 3, is_residual=True)
        x = self.encoderProcessingBlock(x, 16)
        x = self.encoderProcessingBlock(x, 16, reduction_factor=2, is_residual=True)
        x = self.encoderProcessingBlock(x, 32)
        x = self.encoderProcessingBlock(x, 32, reduction_factor=2, is_residual=True)
        x = self.encoderProcessingBlock(x, 64)
        x = self.encoderProcessingBlock(x, 64, reduction_factor=2, is_residual=True)
        x = self.encoderProcessingBlock(x, 128, reduction_factor=2)
        x = Flatten()(x)
        x = Dense(16*16,activation="tanh")(x)
        # Reparametrization trick
        self.encoded_mean = Dense(latent_size)(x)
        self.encoded_sig = Dense(latent_size)(x)
        self.encoded_Data = Lambda(self.sampling, output_shape=(latent_size,), name="Encoded")([self.encoded_mean, self.encoded_sig])

        # Decoder
        latent_input=Input((latent_size))
        x = Dense(16*16,activation="tanh")(latent_input)
        x = Reshape((16,16,1))(x)
        x = self.decoderProcessingBlock(x, 128)
        x = self.decoderProcessingBlock(x, 128, is_residual=True)
        x = self.decoderProcessingBlock(x, 64)
        x = self.decoderProcessingBlock(x, 64, upscaling_factor=2, is_residual=True)
        x = self.decoderProcessingBlock(x, 32)
        x = self.decoderProcessingBlock(x, 32, upscaling_factor=2, is_residual=True)
        x = self.decoderProcessingBlock(x, 16)
        x = self.decoderProcessingBlock(x, 16, upscaling_factor=2, is_residual=True)
        x = self.encoderProcessingBlock(x, 3)
        x = self.encoderProcessingBlock(x, 3, is_residual=True)
        decoded = self.decoderProcessingBlock(x, 3)
        

        # Build models
        self.encoder=Model(input_data,[self.encoded_mean,self.encoded_sig,self.encoded_Data], name="encoder")
        self.decoder=Model(latent_input,decoded, name="decoder")
        self.bvae=Model(input_data,self.decoder(self.encoder(input_data)[2]), name="bvae")

        # Compile the models and show some summaries
        self.bvae.compile(loss=self.loss)

        # Show the three main components the encoder, the decoder and the full VAE
        self.encoder.summary()
        self.decoder.summary()
        self.bvae.summary()
    

"""
        # Build encoder
        input_data = Input(image_shape,name="input")
        x = self.encoderProcessingBlock(input_data, 64)
        x = self.encoderProcessingBlock(x, 128)
        x = Flatten()(x)
        x = Dense(16*16,activation="tanh")(x)
        # Reparametrization trick
        self.encoded_mean = Dense(latent_size)(x)
        self.encoded_sig = Dense(latent_size)(x)
        self.encoded_Data = Lambda(self.sampling, output_shape=(latent_size,), name="Encoded")([self.encoded_mean, self.encoded_sig])
        
        # Decoder
        latent_input=Input((latent_size))
        x = Dense(16*16,activation="tanh")(latent_input)
        x = Reshape((16,16,1))(x)
        x = self.decoderProcessingBlock(x,32, upscaling_factor=2)
        x_r = self.decoderProcessingBlock(x,64, upscaling_factor=2)
        x_g = self.decoderProcessingBlock(x,64, upscaling_factor=2)
        x_b = self.decoderProcessingBlock(x,64, upscaling_factor=2)

        x_r = self.decoderProcessingBlock(x_r,128, upscaling_factor=2)
        x_g = self.decoderProcessingBlock(x_g,128, upscaling_factor=2)
        x_b = self.decoderProcessingBlock(x_b,128, upscaling_factor=2)
        
        x_r = self.decoderProcessingBlock(x_r,1, upscaling_factor=1)
        x_g = self.decoderProcessingBlock(x_g,1, upscaling_factor=1)
        x_b = self.decoderProcessingBlock(x_b,1, upscaling_factor=1)
        decoded = Concatenate(axis=3)([x_r,x_g,x_b])

"""
